src/nlp/fingerprint.py

```diff
@@ -0,0 +1,164 @@
+        type(use_kwargs)
+    )
+    assert ignore_kwargs is None or isinstance(
+        ignore_kwargs, list
+    ), "ignore_kwargs is supposed to be a list, not {}".format(type(use_kwargs))
+    assert not inplace or not fingerprint_names, "fingerprint_names are only used when inplace is False"
+    fingerprint_names = fingerprint_names if fingerprint_names is not None else ["new_fingerprint"]
+
+    def _fingerprint(func):
+
+        assert inplace or all(
+            name in func.__code__.co_varnames for name in fingerprint_names
+        ), "function {} is missing parameters {} in signature".format(func, fingerprint_names)
+        if randomized_function:  # randomized function have seed and generator parameters
+            assert "seed" in func.__code__.co_varnames, "'seed' must be in {}'s signature".format(func)
+            assert "generator" in func.__code__.co_varnames, "'generator' must be in {}'s signature".format(func)
+
+        @wraps(func)
+        def wrapper(*args, **kwargs):
+            self: "Dataset" = args[0]
+            args = args[1:]
+            kwargs_for_fingerprint = dict(kwargs)
+            kwargs_for_fingerprint.update(zip(func.__code__.co_varnames, args))
+
+            # keep the right kwargs to be hashed to generate the fingerprint
+
+            if use_kwargs:
+                kwargs_for_fingerprint = {k: v for k, v in kwargs_for_fingerprint.items() if k in use_kwargs}
+            if ignore_kwargs:
+                kwargs_for_fingerprint = {k: v for k, v in kwargs_for_fingerprint.items() if k not in ignore_kwargs}
+            if randomized_function:  # randomized functions have `seed` and `generator` parameters
+                if kwargs_for_fingerprint.get("seed") is None and kwargs_for_fingerprint.get("generator") is None:
+                    kwargs_for_fingerprint["generator"] = np.random.default_rng(None)
+
+            # add new_fingerprint arg to not in-place transforms
+
+            if not inplace:
+                for fingerprint_name in fingerprint_names:  # transforms like `train_test_split` have several hashes
+                    if fingerprint_name not in kwargs:
+                        kwargs_for_fingerprint["fingerprint_name"] = fingerprint_name
+                        kwargs[fingerprint_name] = update_fingerprint(self._fingerprint, func, kwargs_for_fingerprint)
+
+            # Call actual function
+
+            out = func(self, *args, **kwargs)
+
+            # Update fingerprint of in-place transforms + update in-place history of transforms
+
+            if inplace:  # update after calling func so that the fingerprint doesn't change if the function fails
+                self._fingerprint = update_fingerprint(self._fingerprint, func, kwargs_for_fingerprint)
+                for inplace_hist_per_file in self._inplace_history:
+                    inplace_hist_per_file["transforms"].append((func.__name__, args, kwargs))
```

Comment:
Same as I mentioned earlier, I think I would update the fingerprint before calling the function in case some of the inner working in `func` do in-place changes on some of the args/kwargs then we might loose the original calling args. What do you think?

Reply:
Or does the fact that we have already un updated fingerprint risk having some side effects in `func`?

Reply:
Ok I see, I will compute the new fingerprint and new history before calling `func`, and then update `self._fingerprint` and `self._inplace_history`

Reply:
I may need to `deepcopy` the args/kwargs that are saved in the history

Reply:
Done. I ended up doing it using deep copy for the args/kwargs stored in the history

I'm merging the PR